pkg/controller/pravegacluster/upgrade.go

```diff
@@ -378,3 +425,18 @@ func (r *ReconcilePravegaCluster) getPodsWithVersion(sts *appsv1.StatefulSet, ve
 	}
 	return pods, nil
 }
+
+func checkUpgradeCondition(p *pravegav1alpha1.PravegaCluster, reason string, updatedReplicas int32) error {
```

Comment:
I think we should check the delay at pod level instead of checking at sts level.
If there are many replicas in the sts and each one takes a little extra delay we could easily run out of our threshold.

Reply:
Threshold should be at pod level, assuming it would never take more than 10 mins to upgrade/deploy a single pod, if we realize it is taking us longer than that, we bail out... and fail the upgrade.

Reply:
The sts will help us managing pods. If a pod is updated, it will show up in the status of sts and this method will update the status in Pravega, see the last line. The `SetUpgradingConditionTrue` will update the upgrading condition with a message showing the number of updated replicas, if that number of updated replicas change, the upgrade condition field `updateTimestamp` will be updated to `now`. So it is not calculating the finishing time of the whole sts, the time is updated when the updated replicas change.